kv/kvserver: TestPromoteNonVoterInAddVoter failed

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17614765?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17614765?buildTab=artifacts#/) on master @ [77d7d8d882fa3959c4ade43f0f1358b76b294666](https://github.com/cockroachdb/cockroach/commits/77d7d8d882fa3959c4ade43f0f1358b76b294666):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestPromoteNonVoterInAddVoter2720334201
    test_log_scope.go:76: use -show-logs to present logs inline
    replicate_queue_test.go:2248: condition failed to evaluate within 45s: from replicate_queue_test.go:2252: expected 3 voters; got 7
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestPromoteNonVoterInAddVoter2720334201
--- FAIL: TestPromoteNonVoterInAddVoter (75.00s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #132827 kv/kvserver: TestPromoteNonVoterInAddVoter failed [A-testing C-bug C-test-failure O-robot P-3 T-kv branch-release-24.1]
- #132564 kv/kvserver: TestPromoteNonVoterInAddVoter failed [C-bug C-test-failure O-robot P-3 T-kv branch-release-24.1.6-rc]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44042

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -2140,7 +2140,8 @@ func iterateOverAllStores(
 // the range log where the added replica type is a LEARNER.
 func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
+	scope := log.Scope(t)
+	defer scope.Close(t)
 
 	// This test is slow under stress/race and can time out when upreplicating /
 	// rebalancing to ensure all stores have the same range count initially, due
@@ -2148,6 +2149,8 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	skip.UnderStress(t)
 	skip.UnderDeadlock(t)
 	skip.UnderRace(t)
+
+	defer testutils.StartExecTrace(t, scope.GetDirectory()).Finish(t)
 
 	ctx := context.Background()
 

--- a/pkg/testutils/BUILD.bazel
+++ b/pkg/testutils/BUILD.bazel
@@ -6,6 +6,7 @@ go_library(
         "backup.go",
         "dir.go",
         "error.go",
+        "exectrace.go",
         "files.go",
         "hook.go",
         "keys.go",
@@ -33,6 +34,7 @@ go_library(
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_dataexmachina_dev_side_eye_go//sideeye",
+        "@com_github_petermattis_goid//:goid",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/testutils/exectrace.go
+++ b/pkg/testutils/exectrace.go
@@ -1,0 +1,71 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package testutils
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime/trace"
+
+	"github.com/petermattis/goid"
+)
+
+type ActiveExecTrace struct {
+	name string
+	file *os.File
+	reg  *trace.Region
+}
+
+// Finish stops the ongoing execution trace, if there is one, and closes the
+// file. It must be called only once.
+func (a *ActiveExecTrace) Finish(t interface {
+	Failed() bool
+	Logf(string, ...interface{})
+}) {
+	if a == nil {
+		return
+	}
+	a.reg.End()
+	trace.Stop()
+	_ = a.file.Close()
+	if !t.Failed() {
+		_ = os.Remove(a.file.Name())
+	} else {
+		t.Logf("execution trace written to %s", a.file.Name())
+	}
+}
+
+// StartExecTrace starts a Go execution trace and returns a handle that allows
+// stopping it. If a trace cannot be started, this is logged and nil is returned.
+// It is valid to stop a nil ActiveExecTrace.
+//
+// This helper is intended to instrument tests for which an execution trace is
+// desired on the next failure.
+func StartExecTrace(
+	t interface {
+		Name() string
+		Logf(string, ...interface{})
+	}, dir string,
+) *ActiveExecTrace {
+	path := filepath.Join(dir, fmt.Sprintf("exectrace_goid_%d.bin", goid.Get()))
+	f, err := os.Create(path)
+	if err != nil {
+		t.Logf("could not create file for execution trace: %s", err)
+		return nil
+	}
+	if err := trace.Start(f); err != nil {
+		_ = f.Close()
+		t.Logf("could not start execution trace: %s", err)
+		return nil
+	}
+	return &ActiveExecTrace{
+		name: t.Name(),
+		file: f,
+		reg:  trace.StartRegion(context.Background(), t.Name()),
+	}
+}

